File: pkg/cli/validate/books.go
/* Copyright 2025 Dnote Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package validate

import (
	"strings"

	"github.com/dnote/dnote/pkg/cli/utils"
	"github.com/pkg/errors"
)

var reservedBookNames = []string{"trash", "conflicts"}

// ErrBookNameReserved is an error incidating that the specified book name is reserved
var ErrBookNameReserved = errors.New("The book name is reserved")

// ErrBookNameNumeric is an error for a book name that only contains numbers
var ErrBookNameNumeric = errors.New("The book name cannot contain only numbers")

// ErrBookNameHasSpace is an error for a book name that has any space
var ErrBookNameHasSpace = errors.New("The book name cannot contain spaces")

// ErrBookNameEmpty is an error for an empty book name
var ErrBookNameEmpty = errors.New("The book name is empty")

// ErrBookNameMultiline is an error for a book name that has linebreaks
var ErrBookNameMultiline = errors.New("The book name contains multiple lines")

func isReservedName(name string) bool {
	for _, n := range reservedBookNames {
		if name == n {
			return true
		}
	}

	return false
}

// BookName validates a book name
func BookName(name string) error {
	if name == "" {
		return ErrBookNameEmpty
	}

	if isReservedName(name) {
		return ErrBookNameReserved
	}

	if utils.IsNumber(name) {
		return ErrBookNameNumeric
	}

	if strings.Contains(name, " ") {
		return ErrBookNameHasSpace
	}

	if strings.Contains(name, "\n") || strings.Contains(name, "\r\n") {
		return ErrBookNameMultiline
	}

	return nil
}

File: jslib/src/helpers/books.ts
/* Copyright (C) 2019 Monomax Software Pty Ltd
 *
 * This file is part of Dnote.
 *
 * Dnote is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Dnote is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Dnote.  If not, see <https://www.gnu.org/licenses/>.
 */

import { BookData } from '../operations/types';

// errBookNameNumeric is an error for book names that only contain numbers
export const errBookNameNumeric = new Error(
  'The book name cannot contain only numbers'
);

// errBookNameHasSpace is an error for book names that have any space
export const errBookNameHasSpace = new Error(
  'The book name cannot contain spaces'
);

// errBookNameHasComma is an error for book names that have any comma
export const errBookNameHasComma = new Error('The book name has comma');

// errBookNameReserved is an error incidating that the specified book name is reserved
export const errBookNameReserved = new Error('The book name is reserved');

const numberRegex = /^\d+$/;

const reservedBookNames = ['trash', 'conflicts'];

// validateBookName validates the given book name and throws error if not valid
export function validateBookName(bookName: string) {
  if (reservedBookNames.indexOf(bookName) > -1) {
    throw errBookNameReserved;
  }

  if (numberRegex.test(bookName)) {
    throw errBookNameNumeric;
  }

  if (bookName.indexOf(' ') > -1) {
    throw errBookNameHasSpace;
  }

  if (bookName.indexOf(',') > -1) {
    throw errBookNameHasComma;
  }
}

// checkDuplicate checks if the given book name has a duplicate in the given array
// of books
export function checkDuplicate(books: BookData[], bookName: string): boolean {
  for (let i = 0; i < books.length; i++) {
    const book = books[i];

    if (book.label === bookName) {
      return true;
    }
  }

  return false;
}
